File: yep/src/main/java/catchla/yep/loader/DiscoverUsersLoader.java
package catchla.yep.loader;

import android.accounts.Account;
import android.content.Context;
import android.support.annotation.NonNull;

import java.util.ArrayList;
import java.util.List;

import catchla.yep.Constants;
import catchla.yep.model.DiscoverQuery;
import catchla.yep.model.Paging;
import catchla.yep.model.User;
import catchla.yep.util.YepAPI;
import catchla.yep.util.YepException;

/**
 * Created by mariotaku on 15/5/27.
 */
public class DiscoverUsersLoader extends CachedYepListLoader<User> implements Constants {

    private final DiscoverQuery mQuery;
    private final Paging mPaging;

    public DiscoverUsersLoader(Context context, Account account, DiscoverQuery query, List<User> oldData,
                               Paging paging, boolean readCache, boolean writeCache) {
        super(context, account, User.class, oldData, readCache, writeCache);
        mQuery = query;
        mPaging = paging;
    }

    @NonNull
    @Override
    protected String getCacheFileName() {
        return "discover_cache_" + getAccount().name;
    }

    @Override
    protected List<User> requestData(final YepAPI yep, List<User> oldData) throws YepException {
        final List<User> list = new ArrayList<>();
        if (oldData != null) {
            list.addAll(oldData);
        }
        for (User topic : yep.getDiscover(mQuery, mPaging)) {
            list.remove(topic);
            list.add(topic);
        }
        return list;
    }


}

File: src/main/java/catchla/yep/loader/DiscoverUsersLoader.kt
package catchla.yep.loader

import android.accounts.Account
import android.content.Context
import catchla.yep.Constants
import catchla.yep.model.DiscoverQuery
import catchla.yep.model.Paging
import catchla.yep.model.User
import catchla.yep.model.YepException
import catchla.yep.util.YepAPI
import java.util.*

/**
 * Created by mariotaku on 15/5/27.
 */
class DiscoverUsersLoader(
        context: Context, account: Account,
        private val query: DiscoverQuery,
        oldData: List<User>,
        private val paging: Paging,
        readCache: Boolean,
        writeCache: Boolean
) : CachedYepListLoader<User>(context, account, User::class.java, oldData, readCache, writeCache), Constants {

    override val cacheFileName: String
        get() = "discover_cache_${account.name}"

    @Throws(YepException::class)
    override fun requestData(yep: YepAPI, oldData: List<User>?): List<User> {
        val list = ArrayList<User>()
        if (oldData != null) {
            list.addAll(oldData)
        }
        for (topic in yep.getDiscover(query, paging)) {
            list.remove(topic)
            list.add(topic)
        }
        return list
    }


}
